File: src/com/prey/backwardcompatibility/FroyoSupport.java
/*******************************************************************************
 * Created by Carlos Yaconi
 * Copyright 2012 Fork Ltd. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.backwardcompatibility;


import android.annotation.TargetApi;
import android.app.admin.DevicePolicyManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.telephony.TelephonyManager;

import com.prey.PreyConfig;
import com.prey.PreyLogger;
import com.prey.receivers.PreyDeviceAdmin;

@TargetApi(Build.VERSION_CODES.FROYO)
public class FroyoSupport {

	private static FroyoSupport _instance;
	private Context ctx;
	private DevicePolicyManager policyManager;
	ComponentName deviceAdmin;
	
	public static FroyoSupport getInstance(Context context){
		if (_instance == null){
			_instance = new FroyoSupport(context);
		}
		return _instance;
	}

	private FroyoSupport(Context context){
		policyManager = (DevicePolicyManager)context.getSystemService(Context.DEVICE_POLICY_SERVICE);
		deviceAdmin = new ComponentName(context, PreyDeviceAdmin.class);
		ctx=context;
	}
	
	public void changePasswordAndLock(String newPass, boolean lock){
	    try {
			if (isAdminActive()) {
				boolean resultLocK=policyManager.resetPassword(newPass,DevicePolicyManager.RESET_PASSWORD_REQUIRE_ENTRY);
				PreyLogger.i("resultLocK:"+resultLocK);
				if (lock)
					lockNow();
			}
		} catch (Exception e) {
			PreyLogger.e("This device couldn't be locked. Honeycomb bug?", e);
		}
	}

	 
	public void lockNow(){
		if (isAdminActive())
			policyManager.lockNow();
	}
	
	public boolean isAdminActive(){
		return policyManager.isAdminActive(deviceAdmin);
	}
	
	public void removeAdminPrivileges(){
		policyManager.removeActiveAdmin(deviceAdmin);
	}
	
	public Intent getAskForAdminPrivilegesIntent(){
		PreyConfig.getPreyConfig(ctx).setSecurityPrivilegesAlreadyPrompted(true);
		Intent intent = new Intent(DevicePolicyManager.ACTION_ADD_DEVICE_ADMIN);
		intent.putExtra(DevicePolicyManager.EXTRA_DEVICE_ADMIN,deviceAdmin);
		intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET);
		//intent.putExtra(DevicePolicyManager.EXTRA_ADD_EXPLANATION, getText(R.string.device_admin_prompt));
		return intent;
	}
	
    public void wipe() {
        if (isAdminActive())
                policyManager.wipeData(0);
    }
    
    public static boolean supportSMS(Context ctx){
		TelephonyManager telephonyManager1 = (TelephonyManager)ctx.getSystemService(Context.TELEPHONY_SERVICE);
    	boolean isPhone = !(telephonyManager1.getPhoneType()==TelephonyManager.PHONE_TYPE_NONE);
		return isPhone ;
    }
	
}

File: app/src/main/java/com/prey/backwardcompatibility/FroyoSupport.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.backwardcompatibility

import android.app.admin.DevicePolicyManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.Build
import android.provider.Settings
import com.prey.exceptions.PreyException
import com.prey.PreyConfig
import com.prey.PreyLogger
import com.prey.PreyUtils
import com.prey.actions.lock.LockAction
import com.prey.receivers.PreyDeviceAdminReceiver

class FroyoSupport private constructor(val context: Context) {

    private val policyManager =
        context.getSystemService(Context.DEVICE_POLICY_SERVICE) as DevicePolicyManager
    var deviceAdmin: ComponentName = ComponentName(context, PreyDeviceAdminReceiver::class.java)

    @Throws(PreyException::class)
    fun changePasswordAndLock(newPass: String?, lock: Boolean) {
        try {
            PreyLogger.d("change0")
            if (isAdminActive()) {
                PreyLogger.d("change1")
                val isPatternSet = LockAction().isPatternSet(context)
                val isPassOrPinSet = LockAction().isPassOrPinSet(context)
                if (!isPatternSet && !isPassOrPinSet) {
                    try {
                        var length = 0
                        if (newPass != null) {
                            length = newPass.length
                        }
                        if ("" == newPass) Settings.System.putInt(
                            context.contentResolver,
                            Settings.System.LOCK_PATTERN_ENABLED,
                            0
                        )
                        if (length >= 4) {
                            policyManager.setPasswordMinimumLength(deviceAdmin, 0)
                            policyManager.setPasswordQuality(
                                deviceAdmin,
                                DevicePolicyManager.PASSWORD_QUALITY_UNSPECIFIED
                            )
                            policyManager.resetPassword(
                                newPass,
                                DevicePolicyManager.RESET_PASSWORD_REQUIRE_ENTRY
                            )
                        } else {
                            if (lock) {
                                lockNow()
                            }
                        }
                    } catch (e1: Exception) {
                        if (lock) {
                            lockNow()
                        }
                        PreyLogger.e("locked: ${e1.message}", e1)
                        throw PreyException("This device couldn't be locked")
                    }
                }
                if (lock) {
                    lockNow()
                }
            }
        } catch (e: Exception) {
            throw PreyException("This device couldn't be locked")
        }
    }

    fun lockNow() {
        if (isAdminActive()) policyManager.lockNow()
    }

    fun isAdminActive(): Boolean = if (!PreyUtils.isChromebook(context)) {
        policyManager.isAdminActive(deviceAdmin)
    } else {
        true
    }

    fun removeAdminPrivileges() {
        policyManager.removeActiveAdmin(deviceAdmin)
    }

    fun getAskForAdminPrivilegesIntent(): Intent {
        PreyConfig.getInstance(context).setSecurityPrivilegesAlreadyPrompted(true)
        val intent = Intent(DevicePolicyManager.ACTION_ADD_DEVICE_ADMIN)
        intent.putExtra(DevicePolicyManager.EXTRA_DEVICE_ADMIN, deviceAdmin)
        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET)
        return intent
    }

    fun wipeData() {
        if (isAdminActive()) policyManager.wipeData(0)
    }

    /**
     * Retrieves the enrollment-specific ID for the device, if available.
     *
     * This method checks if the device is running Android S (API level 31) or later,
     * and attempts to retrieve the enrollment-specific ID using the DevicePolicyManager.
     *
     * @return The enrollment-specific ID, or an empty string if not available.
     */
    fun getEnrollmentSpecificId(): String {
        // Initialize the ID as an empty string
        var id = ""
        // Check if the device is running Android S (API level 31) or later
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            try {
                // Retrieve the organization ID from the PreyConfig
                val organizationId = PreyConfig.getInstance(context).getOrganizationId()
                // Check if the organization ID is not null and not empty
                if (organizationId != null && "" != organizationId) {
                    // Set the organization ID before attempting to retrieve the enrollment-specific ID
                    policyManager.setOrganizationId(
                        PreyConfig.getInstance(context).getOrganizationId()!!
                    )
                    // Attempt to retrieve the enrollment-specific ID using the DevicePolicyManager
                    id = policyManager.enrollmentSpecificId
                }
            } catch (e: Exception) {
                // Log any exceptions that occur during the retrieval process
                PreyLogger.e("Failed to get enrollment specific ID", e)
            }
        }
        // Return the retrieved ID, or an empty string if not available
        return id
    }

    companion object {
        private var instance: FroyoSupport? = null
        fun getInstance(context: Context): FroyoSupport =
            instance ?: FroyoSupport(context).also { instance = it }
    }

}